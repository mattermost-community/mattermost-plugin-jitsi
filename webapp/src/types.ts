import {ReactNode} from 'react';

import {GlobalState as ReduxGlobalState} from 'mattermost-redux/types/store';
import {Post} from 'mattermost-redux/types/posts';

export type Config = {
    TeammateNameDisplay?: string
}

export type GlobalState = ReduxGlobalState & {
    'plugins-jitsi': {
        openMeeting: Post | null,
        openMeetingJwt: string | null,
        config: {
            embedded?: boolean,
            /* eslint-disable camelcase */
            naming_scheme?: 'ask' | 'words' | 'mattermost' | 'uuid',
            use_jaas?: boolean
            show_prejoin_page: boolean
            /* eslint-enable camelcase */
        }
    }
}

export type InputElementType = HTMLInputElement | HTMLTextAreaElement;

export type RadioOptionsType = {
    value: string;
    checked: boolean;
    label: string | ReactNode;
}

export enum InputTypes {
    Text = 'text',
    Number = 'number',
    TextArea = 'textarea',
}

export type plugin = 'plugins-jitsi'
