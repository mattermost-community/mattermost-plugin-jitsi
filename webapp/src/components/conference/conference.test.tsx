import * as React from 'react';
import {describe, expect, it, jest} from '@jest/globals';
import {shallow} from 'enzyme';

import {Post, PostMetadata, PostType} from 'mattermost-redux/types/posts';

import Conference from './conference';

describe('Conference', () => {
    const basePost: Post = {
        id: 'test',
        create_at: 100,
        update_at: 100,
        edit_at: 100,
        delete_at: 100,
        message: 'test-message',
        is_pinned: false,
        user_id: 'test-user-id',
        channel_id: 'test-channel-id',
        root_id: '',
        parent_id: '',
        original_id: '',
        type: 'custom_jitsi' as PostType,
        hashtags: '',
        props: {
            jwt_meeting_valid_until: 123,
            meeting_link: 'http://test-meeting-link/test',
            jwt_meeting: true,
            meeting_jwt: 'xxxxxxxxxxxx',
            meeting_topic: 'Test topic',
            meeting_id: 'test_meeting_id',
            meeting_personal: false
        },
        metadata: {} as PostMetadata,
        pending_post_id: 'test',
        reply_count: 100
    };

    const actions = {
        openJitsiMeeting: jest.fn(),
        setUserStatus: jest.fn()
    };

    const defaultProps = {
        post: basePost,
        jwt: null,
        actions,
        currentUser: {
            id: 'mockId',
            username: 'firstLast'
        }
    };

    Conference.prototype.getViewportWidth = () => 10;
    Conference.prototype.getViewportHeight = () => 10;
    Conference.prototype.componentDidUpdate = jest.fn();
    Conference.prototype.componentDidMount = jest.fn();
    Conference.prototype.componentWillUnmount = jest.fn();

    it('should render null if the post type is null', () => {
        const props = {...defaultProps, post: null};
        const wrapper = shallow(
            <Conference {...props}/>
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('should render and initialize the conference interface', () => {
        const wrapper = shallow(
            <Conference {...defaultProps}/>
        );
        expect(wrapper).toMatchSnapshot();
    });

    describe('should show the correct buttons depending on the state', () => {
        const wrapper = shallow(
            <Conference {...defaultProps}/>
        );
        const instance = wrapper.instance();

        it('should have down, open outside, maximize and close buttons', () => {
            instance.setState({minimized: true, position: 'top'});
            expect(wrapper).toMatchSnapshot();
        });

        it('should have up, open outside, maximize and close buttons', () => {
            instance.setState({minimized: true, position: 'bottom'});
            expect(wrapper).toMatchSnapshot();
        });

        it('should have open outside, minimize and close buttons', () => {
            instance.setState({minimized: false});
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('should show the the loading spinner depending on the state', () => {
        const wrapper = shallow(
            <Conference {...defaultProps}/>
        );
        const instance = wrapper.instance();

        it('should show loading', () => {
            instance.setState({loading: true});
            expect(wrapper).toMatchSnapshot();
        });

        it('should not show loading', () => {
            instance.setState({loading: false});
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('should maximize based on the state', () => {
        const wrapper = shallow(
            <Conference {...defaultProps}/>
        );
        const instance = wrapper.instance() as Conference;

        it('should toggle tile only if was open before minimize and now is closed', () => {
            instance.api = {
                executeCommand: jest.fn()
            };
            instance.setState({wasTileView: true, isTileView: true});
            instance.maximize();
            expect(instance.api.executeCommand).not.toBeCalledWith('toggleTileView');
            instance.api.executeCommand.mockClear();

            instance.setState({wasTileView: true, isTileView: false});
            instance.maximize();
            expect(instance.api.executeCommand).toBeCalledWith('toggleTileView');
            instance.api.executeCommand.mockClear();

            instance.setState({wasTileView: false, isTileView: true});
            instance.maximize();
            expect(instance.api.executeCommand).toBeCalledWith('toggleTileView');
            instance.api.executeCommand.mockClear();

            instance.setState({wasTileView: false, isTileView: false});
            instance.maximize();
            expect(instance.api.executeCommand).not.toBeCalledWith('toggleTileView');
            instance.api.executeCommand.mockClear();
        });

        it('should toggle filmstrip only if was open before minimize and now is closed', () => {
            instance.api = {
                executeCommand: jest.fn()
            };
            instance.setState({wasFilmStrip: true, isFilmStrip: true});
            instance.maximize();
            expect(instance.api.executeCommand).not.toBeCalledWith('toggleFilmStrip');
            instance.api.executeCommand.mockClear();

            instance.setState({wasFilmStrip: true, isFilmStrip: false});
            instance.maximize();
            expect(instance.api.executeCommand).toBeCalledWith('toggleFilmStrip');
            instance.api.executeCommand.mockClear();

            instance.setState({wasFilmStrip: false, isFilmStrip: true});
            instance.maximize();
            expect(instance.api.executeCommand).toBeCalledWith('toggleFilmStrip');
            instance.api.executeCommand.mockClear();

            instance.setState({wasFilmStrip: false, isFilmStrip: false});
            instance.maximize();
            expect(instance.api.executeCommand).not.toBeCalledWith('toggleFilmStrip');
            instance.api.executeCommand.mockClear();
        });
    });

    describe('should minimize based on the state', () => {
        const wrapper = shallow(
            <Conference {...defaultProps}/>
        );
        const instance = wrapper.instance() as Conference;

        it('should toggle tile only if is open before minimizing', () => {
            instance.api = {
                executeCommand: jest.fn()
            };
            instance.setState({isTileView: true});
            instance.minimize();
            expect(instance.api.executeCommand).toBeCalledWith('toggleTileView');
            instance.api.executeCommand.mockClear();

            instance.setState({isTileView: false});
            instance.minimize();
            expect(instance.api.executeCommand).not.toBeCalledWith('toggleTileView');
            instance.api.executeCommand.mockClear();
        });

        it('should toggle filmstrip only if is open before minimize', () => {
            instance.api = {
                executeCommand: jest.fn()
            };
            instance.setState({isFilmStrip: true});
            instance.minimize();
            expect(instance.api.executeCommand).toBeCalledWith('toggleFilmStrip');
            instance.api.executeCommand.mockClear();

            instance.setState({isFilmStrip: false});
            instance.minimize();
            expect(instance.api.executeCommand).not.toBeCalledWith('toggleFilmStrip');
            instance.api.executeCommand.mockClear();
        });
    });

    it('should execute the hangup command, wait and call the action to close the meeting, and reset the state on closed', (done) => {
        defaultProps.actions.openJitsiMeeting.mockClear();
        const wrapper = shallow(
            <Conference {...defaultProps}/>
        );
        const instance = wrapper.instance() as Conference;
        instance.setState({
            minimized: false,
            loading: false,
            position: 'top',
            wasTileView: false,
            isTileView: false,
            wasFilmStrip: false,
            isFilmStrip: false
        });
        instance.api = {
            executeCommand: jest.fn(),
            dispose: jest.fn()
        };
        instance.close();
        expect(defaultProps.actions.openJitsiMeeting).not.toBeCalled();
        expect(instance.api.executeCommand).toBeCalledWith('hangup');
        setTimeout(() => {
            expect(defaultProps.actions.openJitsiMeeting).toBeCalledTimes(1);
            expect(defaultProps.actions.openJitsiMeeting).toBeCalledWith(null, null);
            expect(instance.api.dispose).toBeCalled();
            expect(instance.state).toEqual({
                minimized: true,
                loading: true,
                position: 'bottom',
                wasTileView: true,
                isTileView: true,
                wasFilmStrip: true,
                isFilmStrip: true
            });

            if (done) {
                done();
            }
        }, 210);
    });
});
