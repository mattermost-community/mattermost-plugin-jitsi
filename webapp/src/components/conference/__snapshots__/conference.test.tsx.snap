// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Conference should render and initialize the conference interface 1`] = `
<Fragment>
  <div
    id="jitsiMeet"
    style={
      Object {
        "alignItems": "center",
        "background": "#474747",
        "bottom": "8px",
        "display": "flex",
        "height": 288,
        "justifyContent": "center",
        "position": "absolute",
        "right": "8px",
        "top": "",
        "width": 384,
        "zIndex": 1100,
      }
    }
  />
  <div
    style={
      Object {
        "alignItems": "center",
        "background": "#00000077",
        "bottom": "8px",
        "color": "white",
        "display": "flex",
        "fontSize": "48px",
        "height": 288,
        "justifyContent": "center",
        "opacity": 0.85,
        "position": "absolute",
        "right": "8px",
        "top": "",
        "width": 384,
        "zIndex": 1200,
      }
    }
  >
    <i
      className="fa fa-spinner fa-fw fa-pulse spinner"
    />
  </div>
  <div
    style={
      Object {
        "bottom": "270px",
        "color": "white",
        "cursor": "pointer",
        "fontSize": "18px",
        "opacity": 0.85,
        "position": "absolute",
        "right": "10px",
        "top": "",
        "zIndex": 1200,
      }
    }
  >
    <FormattedMessage
      defaultMessage="Move up"
      id="jitsi.move-up"
      values={Object {}}
    >
      <Component />
    </FormattedMessage>
    <FormattedMessage
      defaultMessage="Maximize"
      id="jitsi.maximize"
      values={Object {}}
    >
      <Component />
    </FormattedMessage>
    <a
      href="http://test-meeting-link/test#config.callDisplayName=\\"Test topic\\""
      onClick={[Function]}
      rel="noopener noreferrer"
      style={
        Object {
          "color": "white",
        }
      }
      target="_blank"
    >
      <FormattedMessage
        defaultMessage="Open in new tab"
        id="jitsi.open-in-new-tab"
        values={Object {}}
      >
        <Component />
      </FormattedMessage>
    </a>
    <FormattedMessage
      defaultMessage="Close"
      id="jitsi.close"
      values={Object {}}
    >
      <Component />
    </FormattedMessage>
  </div>
</Fragment>
`;

exports[`Conference should render null if the post type is null 1`] = `""`;

exports[`Conference should show the correct buttons depending on the state should have down, open outside, maximize and close buttons 1`] = `
<Fragment>
  <div
    id="jitsiMeet"
    style={
      Object {
        "alignItems": "center",
        "background": "#474747",
        "bottom": "",
        "display": "flex",
        "height": 288,
        "justifyContent": "center",
        "position": "absolute",
        "right": "8px",
        "top": "8px",
        "width": 384,
        "zIndex": 1100,
      }
    }
  />
  <div
    style={
      Object {
        "alignItems": "center",
        "background": "#00000077",
        "bottom": "",
        "color": "white",
        "display": "flex",
        "fontSize": "48px",
        "height": 288,
        "justifyContent": "center",
        "opacity": 0.85,
        "position": "absolute",
        "right": "8px",
        "top": "8px",
        "width": 384,
        "zIndex": 1200,
      }
    }
  >
    <i
      className="fa fa-spinner fa-fw fa-pulse spinner"
    />
  </div>
  <div
    style={
      Object {
        "bottom": "",
        "color": "white",
        "cursor": "pointer",
        "fontSize": "18px",
        "opacity": 0.85,
        "position": "absolute",
        "right": "10px",
        "top": "8px",
        "zIndex": 1200,
      }
    }
  >
    <FormattedMessage
      defaultMessage="Move down"
      id="jitsi.move-down"
      values={Object {}}
    >
      <Component />
    </FormattedMessage>
    <FormattedMessage
      defaultMessage="Maximize"
      id="jitsi.maximize"
      values={Object {}}
    >
      <Component />
    </FormattedMessage>
    <a
      href="http://test-meeting-link/test#config.callDisplayName=\\"Test topic\\""
      onClick={[Function]}
      rel="noopener noreferrer"
      style={
        Object {
          "color": "white",
        }
      }
      target="_blank"
    >
      <FormattedMessage
        defaultMessage="Open in new tab"
        id="jitsi.open-in-new-tab"
        values={Object {}}
      >
        <Component />
      </FormattedMessage>
    </a>
    <FormattedMessage
      defaultMessage="Close"
      id="jitsi.close"
      values={Object {}}
    >
      <Component />
    </FormattedMessage>
  </div>
</Fragment>
`;

exports[`Conference should show the correct buttons depending on the state should have open outside, minimize and close buttons 1`] = `
<Fragment>
  <div
    id="jitsiMeet"
    style={
      Object {
        "alignItems": "center",
        "background": "#474747",
        "bottom": "8px",
        "display": "flex",
        "height": 10,
        "justifyContent": "center",
        "position": "absolute",
        "right": "8px",
        "top": "",
        "width": 10,
        "zIndex": 1100,
      }
    }
  />
  <div
    style={
      Object {
        "background": "black",
        "height": "100%",
        "left": 0,
        "opacity": 0.7,
        "position": "absolute",
        "top": 0,
        "width": "100%",
        "zIndex": 1000,
      }
    }
  />
  <div
    style={
      Object {
        "alignItems": "center",
        "background": "#00000077",
        "bottom": "8px",
        "color": "white",
        "display": "flex",
        "fontSize": "48px",
        "height": 10,
        "justifyContent": "center",
        "opacity": 0.85,
        "position": "absolute",
        "right": "8px",
        "top": "",
        "width": 10,
        "zIndex": 1200,
      }
    }
  >
    <i
      className="fa fa-spinner fa-fw fa-pulse spinner"
    />
  </div>
  <div
    style={
      Object {
        "bottom": "-8px",
        "color": "white",
        "cursor": "pointer",
        "fontSize": "18px",
        "opacity": 0.85,
        "position": "absolute",
        "right": "10px",
        "top": "",
        "zIndex": 1200,
      }
    }
  >
    <FormattedMessage
      defaultMessage="Minimize"
      id="jitsi.minimize"
      values={Object {}}
    >
      <Component />
    </FormattedMessage>
    <a
      href="http://test-meeting-link/test#config.callDisplayName=\\"Test topic\\""
      onClick={[Function]}
      rel="noopener noreferrer"
      style={
        Object {
          "color": "white",
        }
      }
      target="_blank"
    >
      <FormattedMessage
        defaultMessage="Open in new tab"
        id="jitsi.open-in-new-tab"
        values={Object {}}
      >
        <Component />
      </FormattedMessage>
    </a>
    <FormattedMessage
      defaultMessage="Close"
      id="jitsi.close"
      values={Object {}}
    >
      <Component />
    </FormattedMessage>
  </div>
</Fragment>
`;

exports[`Conference should show the correct buttons depending on the state should have up, open outside, maximize and close buttons 1`] = `
<Fragment>
  <div
    id="jitsiMeet"
    style={
      Object {
        "alignItems": "center",
        "background": "#474747",
        "bottom": "8px",
        "display": "flex",
        "height": 288,
        "justifyContent": "center",
        "position": "absolute",
        "right": "8px",
        "top": "",
        "width": 384,
        "zIndex": 1100,
      }
    }
  />
  <div
    style={
      Object {
        "alignItems": "center",
        "background": "#00000077",
        "bottom": "8px",
        "color": "white",
        "display": "flex",
        "fontSize": "48px",
        "height": 288,
        "justifyContent": "center",
        "opacity": 0.85,
        "position": "absolute",
        "right": "8px",
        "top": "",
        "width": 384,
        "zIndex": 1200,
      }
    }
  >
    <i
      className="fa fa-spinner fa-fw fa-pulse spinner"
    />
  </div>
  <div
    style={
      Object {
        "bottom": "270px",
        "color": "white",
        "cursor": "pointer",
        "fontSize": "18px",
        "opacity": 0.85,
        "position": "absolute",
        "right": "10px",
        "top": "",
        "zIndex": 1200,
      }
    }
  >
    <FormattedMessage
      defaultMessage="Move up"
      id="jitsi.move-up"
      values={Object {}}
    >
      <Component />
    </FormattedMessage>
    <FormattedMessage
      defaultMessage="Maximize"
      id="jitsi.maximize"
      values={Object {}}
    >
      <Component />
    </FormattedMessage>
    <a
      href="http://test-meeting-link/test#config.callDisplayName=\\"Test topic\\""
      onClick={[Function]}
      rel="noopener noreferrer"
      style={
        Object {
          "color": "white",
        }
      }
      target="_blank"
    >
      <FormattedMessage
        defaultMessage="Open in new tab"
        id="jitsi.open-in-new-tab"
        values={Object {}}
      >
        <Component />
      </FormattedMessage>
    </a>
    <FormattedMessage
      defaultMessage="Close"
      id="jitsi.close"
      values={Object {}}
    >
      <Component />
    </FormattedMessage>
  </div>
</Fragment>
`;

exports[`Conference should show the the loading spinner depending on the state should not show loading 1`] = `
<Fragment>
  <div
    id="jitsiMeet"
    style={
      Object {
        "alignItems": "center",
        "background": "#474747",
        "bottom": "8px",
        "display": "flex",
        "height": 288,
        "justifyContent": "center",
        "position": "absolute",
        "right": "8px",
        "top": "",
        "width": 384,
        "zIndex": 1100,
      }
    }
  />
  <div
    style={
      Object {
        "bottom": "270px",
        "color": "white",
        "cursor": "pointer",
        "fontSize": "18px",
        "opacity": 0.85,
        "position": "absolute",
        "right": "10px",
        "top": "",
        "zIndex": 1200,
      }
    }
  >
    <FormattedMessage
      defaultMessage="Move up"
      id="jitsi.move-up"
      values={Object {}}
    >
      <Component />
    </FormattedMessage>
    <FormattedMessage
      defaultMessage="Maximize"
      id="jitsi.maximize"
      values={Object {}}
    >
      <Component />
    </FormattedMessage>
    <a
      href="http://test-meeting-link/test#config.callDisplayName=\\"Test topic\\""
      onClick={[Function]}
      rel="noopener noreferrer"
      style={
        Object {
          "color": "white",
        }
      }
      target="_blank"
    >
      <FormattedMessage
        defaultMessage="Open in new tab"
        id="jitsi.open-in-new-tab"
        values={Object {}}
      >
        <Component />
      </FormattedMessage>
    </a>
    <FormattedMessage
      defaultMessage="Close"
      id="jitsi.close"
      values={Object {}}
    >
      <Component />
    </FormattedMessage>
  </div>
</Fragment>
`;

exports[`Conference should show the the loading spinner depending on the state should show loading 1`] = `
<Fragment>
  <div
    id="jitsiMeet"
    style={
      Object {
        "alignItems": "center",
        "background": "#474747",
        "bottom": "8px",
        "display": "flex",
        "height": 288,
        "justifyContent": "center",
        "position": "absolute",
        "right": "8px",
        "top": "",
        "width": 384,
        "zIndex": 1100,
      }
    }
  />
  <div
    style={
      Object {
        "alignItems": "center",
        "background": "#00000077",
        "bottom": "8px",
        "color": "white",
        "display": "flex",
        "fontSize": "48px",
        "height": 288,
        "justifyContent": "center",
        "opacity": 0.85,
        "position": "absolute",
        "right": "8px",
        "top": "",
        "width": 384,
        "zIndex": 1200,
      }
    }
  >
    <i
      className="fa fa-spinner fa-fw fa-pulse spinner"
    />
  </div>
  <div
    style={
      Object {
        "bottom": "270px",
        "color": "white",
        "cursor": "pointer",
        "fontSize": "18px",
        "opacity": 0.85,
        "position": "absolute",
        "right": "10px",
        "top": "",
        "zIndex": 1200,
      }
    }
  >
    <FormattedMessage
      defaultMessage="Move up"
      id="jitsi.move-up"
      values={Object {}}
    >
      <Component />
    </FormattedMessage>
    <FormattedMessage
      defaultMessage="Maximize"
      id="jitsi.maximize"
      values={Object {}}
    >
      <Component />
    </FormattedMessage>
    <a
      href="http://test-meeting-link/test#config.callDisplayName=\\"Test topic\\""
      onClick={[Function]}
      rel="noopener noreferrer"
      style={
        Object {
          "color": "white",
        }
      }
      target="_blank"
    >
      <FormattedMessage
        defaultMessage="Open in new tab"
        id="jitsi.open-in-new-tab"
        values={Object {}}
      >
        <Component />
      </FormattedMessage>
    </a>
    <FormattedMessage
      defaultMessage="Close"
      id="jitsi.close"
      values={Object {}}
    >
      <Component />
    </FormattedMessage>
  </div>
</Fragment>
`;
